File: dynasty-scout/scrapers/redraft/player_props.py
"""
Turn the player prop market into a fantasy projection.

Every other projection here is an opinion. A prop line is a price — the number
at which a book will take money on both sides — and it reprices on news hours
before any ranking does. Converting a week's lines into PPR points gives a
per-player projection that is, for that week, the sharpest one available.

The conversion is the whole idea and it is deliberately plain:

    PPR = receptions x 1
        + receiving yards x 0.1
        + rushing yards x 0.1
        + passing yards x 0.04
        + passing touchdowns x 4
        + P(scores a touchdown) x 6

A yardage line is a median rather than a mean, and for these distributions the
two are close enough that the difference is far smaller than the disagreement
between books. Touchdowns are the exception and are not taken from a line at
all: the anytime-touchdown price converts to a probability, which is what the
6 points multiply.

Vig comes off first. A book prices both sides to sum past 100%, so each side's
raw probability is scaled by that overround; on a two-sided market that is
exact, and on anytime-touchdown — usually priced yes-only — it falls back to
the raw implied probability, which is a slight overstatement and is recorded
as such rather than silently corrected.

Source: The Odds API (the-odds-api.com), which needs a key in ODDS_API_KEY.
Without one this loader does nothing and says so; nothing downstream treats a
missing market as a zero.

Usage:  py -m scrapers.redraft.player_props [--season 2026] [--week 2]
"""
import argparse
import json
import os
import sys
import urllib.error
import urllib.request
from collections import defaultdict
from datetime import datetime

from scrapers import config
from scrapers.redraft import nflverse_stats as base
from scrapers.redraft.names import normalize_name

API = "https://api.the-odds-api.com/v4"
SPORT = "americanfootball_nfl"

# The Odds API market keys we care about, mapped to our own names.
MARKETS = {
    "player_receptions": "receptions",
    "player_reception_yds": "rec_yds",
    "player_rush_yds": "rush_yds",
    "player_rush_attempts": "rush_attempts",
    "player_pass_yds": "pass_yds",
    "player_pass_tds": "pass_tds",
    "player_anytime_td": "anytime_td",
}

# PPR value of one unit of each market. anytime_td is handled separately
# because it is a probability, not a count.
PPR_PER_UNIT = {
    "receptions": 1.0,
    "rec_yds": 0.1,
    "rush_yds": 0.1,
    "pass_yds": 0.04,
    "pass_tds": 4.0,
    "rush_attempts": 0.0,   # priced for context, worth no points itself
}
TD_POINTS = 6.0


def american_to_prob(price):
    """American odds to their implied probability, vig included."""
    if price is None:
        return None
    p = float(price)
    if p == 0:
        return None
    return (-p) / ((-p) + 100.0) if p < 0 else 100.0 / (p + 100.0)


def devig(over_price, under_price):
    """
    The over's probability with the book's margin removed.

    Both sides priced means the overround is measurable and the split is
    exact. One side only means the best available answer is the raw implied
    probability, which runs a few points high; callers are told which they got
    by whether under_price was there at all.
    """
    po = american_to_prob(over_price)
    pu = american_to_prob(under_price)
    if po is None:
        return None
    if pu is None:
        return po
    total = po + pu
    return po / total if total > 0 else po


def ppr_from_markets(values):
    """
    PPR points from a player's priced markets.

    Returns (points, markets_used). Markets that were not priced contribute
    nothing and are not guessed at — a receiver with only a yardage line gets
    a thinner projection, and markets_used says so.
    """
    pts, used = 0.0, 0
    for market, per_unit in PPR_PER_UNIT.items():
        v = values.get(market)
        if v is None:
            continue
        used += 1
        pts += v * per_unit
    td = values.get("anytime_td_prob")
    if td is not None:
        used += 1
        pts += td * TD_POINTS
    return round(pts, 2), used


def fetch_events(api_key, season, week):
    url = f"{API}/sports/{SPORT}/events?apiKey={api_key}"
    req = urllib.request.Request(url, headers=base.HEADERS)
    with urllib.request.urlopen(req, timeout=60) as r:
        return json.loads(r.read().decode())


def fetch_event_odds(api_key, event_id):
    markets = ",".join(MARKETS)
    url = (f"{API}/sports/{SPORT}/events/{event_id}/odds"
           f"?apiKey={api_key}&regions=us&oddsFormat=american&markets={markets}")
    req = urllib.request.Request(url, headers=base.HEADERS)
    with urllib.request.urlopen(req, timeout=60) as r:
        return json.loads(r.read().decode())


def parse_event(event, by_name):
    """
    One event's odds payload into prop rows.

    Kept free of network and database so the shape this assumes can be pinned
    by a fixture: outcomes carry the player in `description`, the line in
    `point`, and Over/Under (or Yes/No) in `name`.
    """
    rows = []
    for bm in event.get("bookmakers") or []:
        book = bm.get("key") or "?"
        for mk in bm.get("markets") or []:
            market = MARKETS.get(mk.get("key"))
            if not market:
                continue
            sides = defaultdict(dict)
            for o in mk.get("outcomes") or []:
                who = (o.get("description") or "").strip()
                if not who:
                    continue
                side = (o.get("name") or "").strip().lower()
                key = 'over' if side in ('over', 'yes') else 'under' if side in ('under', 'no') else side
                sides[who][key] = o
            for who, s in sides.items():
                pid = by_name.get(normalize_name(who))
                if not pid:
                    continue
                over, under = s.get('over'), s.get('under')
                if not over:
                    continue
                rows.append({
                    "player_id": pid, "market": market, "book": book,
                    "line": over.get("point"),
                    "over_price": over.get("price"),
                    "under_price": (under or {}).get("price"),
                    "over_prob": devig(over.get("price"), (under or {}).get("price")),
                    "event_id": event.get("id"),
                    "commence_time": event.get("commence_time"),
                })
    return rows


def consensus(rows):
    """
    One projection per player from however many books priced them.

    The median across books rather than the mean: books that have not moved a
    stale line should not drag the number, and with a handful of prices the
    median is the robust choice.
    """
    by_player = defaultdict(lambda: defaultdict(list))
    books = defaultdict(set)
    for r in rows:
        books[r["player_id"]].add(r["book"])
        if r["market"] == "anytime_td":
            if r["over_prob"] is not None:
                by_player[r["player_id"]]["anytime_td_prob"].append(r["over_prob"])
        elif r["line"] is not None:
            by_player[r["player_id"]][r["market"]].append(float(r["line"]))

    out = {}
    for pid, markets in by_player.items():
        values = {}
        for m, vals in markets.items():
            vals = sorted(vals)
            mid = len(vals) // 2
            values[m] = vals[mid] if len(vals) % 2 else (vals[mid - 1] + vals[mid]) / 2
        pts, used = ppr_from_markets(values)
        out[pid] = {**values, "ppr_points": pts, "markets_priced": used,
                    "books_priced": len(books[pid])}
    return out


PROP_INSERT = """
INSERT OR REPLACE INTO nfl_player_prop (
    player_id, season, week, market, book, line, over_price, under_price,
    over_prob, event_id, commence_time, scraped_at, updated_at
) VALUES (?,?,?,?,?,?,?,?,?,?,?,?,?)
"""

PROJ_INSERT = """
INSERT OR REPLACE INTO nfl_player_market_projection (
    player_id, season, week, receptions, rec_yards, rush_yards, pass_yards,
    pass_tds, rush_attempts, anytime_td_prob, ppr_points, markets_priced,
    books_priced, scraped_at, updated_at
) VALUES (?,?,?,?,?,?,?,?,?,?,?,?,?,?,?)
"""


def load_name_index(cursor):
    """Normalised name -> player id, for the unambiguous names only."""
    cursor.execute(
        "SELECT id, full_name FROM players WHERE redraft_pool = 1 AND position != 'DST'")
    seen = defaultdict(list)
    for r in cursor.fetchall():
        seen[normalize_name(r["full_name"])].append(r["id"])
    return {k: v[0] for k, v in seen.items() if len(v) == 1}


def persist(cursor, season, week, rows, proj, scraped_at, now):
    cursor.executemany(PROP_INSERT, [(
        r["player_id"], season, week, r["market"], r["book"], r["line"],
        r["over_price"], r["under_price"], r["over_prob"], r["event_id"],
        r["commence_time"], scraped_at, now,
    ) for r in rows])
    cursor.executemany(PROJ_INSERT, [(
        pid, season, week, v.get("receptions"), v.get("rec_yds"),
        v.get("rush_yds"), v.get("pass_yds"), v.get("pass_tds"),
        v.get("rush_attempts"), v.get("anytime_td_prob"), v["ppr_points"],
        v["markets_priced"], v["books_priced"], scraped_at, now,
    ) for pid, v in proj.items()])


def run(season, week):
    api_key = os.environ.get("ODDS_API_KEY", "").strip()
    if not api_key:
        print("ODDS_API_KEY is not set — no props loaded.")
        print("The model treats missing props as missing, not as zero, so the")
        print("rest of the pipeline is unaffected. Set the key to turn this on.")
        return 1

    conn = config.get_db_connection()
    cur = conn.cursor()
    by_name = load_name_index(cur)
    scraped_at = datetime.now().strftime("%Y-%m-%d")
    now = datetime.now().isoformat(timespec="seconds")

    try:
        events = fetch_events(api_key, season, week)
    except urllib.error.HTTPError as e:
        print(f"events request failed: {e.code} {e.reason}")
        return 2

    rows = []
    for ev in events:
        try:
            odds = fetch_event_odds(api_key, ev["id"])
        except urllib.error.HTTPError as e:
            print(f"  {ev.get('id')}: {e.code} {e.reason}")
            continue
        rows.extend(parse_event(odds, by_name))

    proj = consensus(rows)
    persist(cur, season, week, rows, proj, scraped_at, now)
    conn.commit()
    print(f"events {len(events)}  prop rows {len(rows)}  players projected {len(proj)}")
    top = sorted(proj.items(), key=lambda kv: -kv[1]["ppr_points"])[:5]
    for pid, v in top:
        name = cur.execute("SELECT full_name FROM players WHERE id = ?", (pid,)).fetchone()
        print(f"  {name['full_name']:24s} {v['ppr_points']:5.1f} PPR "
              f"({v['markets_priced']} markets, {v['books_priced']} books)")
    conn.close()
    return 0


if __name__ == "__main__":
    p = argparse.ArgumentParser()
    p.add_argument("--season", type=int, default=2026)
    p.add_argument("--week", type=int, required=True)
    a = p.parse_args()
    sys.exit(run(a.season, a.week))

File: dynasty-scout/lib/playerProps.ts
/**
 * The player prop market, as a fantasy projection, refreshed from the server.
 *
 * A prop line is a price rather than an opinion, and it moves on news hours
 * before any ranking does — which only helps if it is re-read often. This is
 * the daily-pass twin of scrapers/redraft/player_props.py, held to the same
 * arithmetic; scripts/props_check.py and the TypeScript checks below run the
 * same fixture so the two cannot quietly drift apart.
 *
 * Needs ODDS_API_KEY. Without one this does nothing and says so: a missing
 * market stays missing, and every player keeps the model's own number.
 */
import { normalizeName } from '@/lib/espnPlayers';
import { getDb, query } from '@/lib/db';

const API = 'https://api.the-odds-api.com/v4';
const SPORT = 'americanfootball_nfl';

/** The Odds API market keys, mapped to ours. */
export const MARKETS: Record<string, string> = {
    player_receptions: 'receptions',
    player_reception_yds: 'rec_yds',
    player_rush_yds: 'rush_yds',
    player_rush_attempts: 'rush_attempts',
    player_pass_yds: 'pass_yds',
    player_pass_tds: 'pass_tds',
    player_anytime_td: 'anytime_td',
};

/** PPR value of one unit. Touchdowns are a probability, handled separately. */
const PPR_PER_UNIT: Record<string, number> = {
    receptions: 1, rec_yds: 0.1, rush_yds: 0.1,
    pass_yds: 0.04, pass_tds: 4, rush_attempts: 0,
};
const TD_POINTS = 6;

/** American odds to their implied probability, vig included. */
export function americanToProb(price: number | null | undefined): number | null {
    if (price == null || price === 0) return null;
    return price < 0 ? -price / (-price + 100) : 100 / (price + 100);
}

/**
 * The over's probability with the book's margin removed.
 *
 * Both sides priced makes the overround measurable and the split exact. One
 * side only leaves the raw implied probability, which runs high; callers can
 * tell which they got from whether the under was there.
 */
export function devig(over: number | null | undefined,
                      under: number | null | undefined): number | null {
    const po = americanToProb(over);
    if (po == null) return null;
    const pu = americanToProb(under);
    if (pu == null) return po;
    const total = po + pu;
    return total > 0 ? po / total : po;
}

/** PPR points from a player's priced markets, and how many were priced. */
export function pprFromMarkets(v: Record<string, number>): [number, number] {
    let pts = 0, used = 0;
    for (const [m, per] of Object.entries(PPR_PER_UNIT)) {
        if (v[m] == null) continue;
        used++; pts += v[m] * per;
    }
    if (v.anytime_td_prob != null) { used++; pts += v.anytime_td_prob * TD_POINTS; }
    return [Math.round(pts * 100) / 100, used];
}

export interface PropRow {
    playerId: number; market: string; book: string;
    line: number | null; overPrice: number | null; underPrice: number | null;
    overProb: number | null; eventId: string | null; commenceTime: string | null;
}

/**
 * One event's odds payload into prop rows.
 *
 * Free of network and database so the shape it assumes can be pinned: the
 * player is in `description`, the line in `point`, and Over/Under (or Yes/No)
 * in `name`.
 */
export function parseEvent(event: any, byName: Map<string, number>): PropRow[] {
    const out: PropRow[] = [];
    for (const bm of event?.bookmakers ?? []) {
        const book = bm?.key ?? '?';
        for (const mk of bm?.markets ?? []) {
            const market = MARKETS[mk?.key];
            if (!market) continue;
            const sides = new Map<string, { over?: any; under?: any }>();
            for (const o of mk?.outcomes ?? []) {
                const who = (o?.description ?? '').trim();
                if (!who) continue;
                const name = (o?.name ?? '').trim().toLowerCase();
                const key = (name === 'over' || name === 'yes') ? 'over'
                    : (name === 'under' || name === 'no') ? 'under' : null;
                if (!key) continue;
                const cur = sides.get(who) ?? {};
                cur[key] = o;
                sides.set(who, cur);
            }
            for (const [who, s] of sides) {
                const pid = byName.get(normalizeName(who));
                if (!pid || !s.over) continue;
                out.push({
                    playerId: pid, market, book,
                    line: s.over.point ?? null,
                    overPrice: s.over.price ?? null,
                    underPrice: s.under?.price ?? null,
                    overProb: devig(s.over.price, s.under?.price),
                    eventId: event?.id ?? null,
                    commenceTime: event?.commence_time ?? null,
                });
            }
        }
    }
    return out;
}

export interface MarketProjection {
    values: Record<string, number>;
    pprPoints: number; marketsPriced: number; booksPriced: number;
}

/**
 * One projection per player from however many books priced them.
 *
 * The median across books, not the mean: a book that has not moved a stale
 * line should not drag the number, and with a handful of prices the median is
 * the robust choice.
 */
export function consensus(rows: PropRow[]): Map<number, MarketProjection> {
    const byPlayer = new Map<number, Map<string, number[]>>();
    const books = new Map<number, Set<string>>();
    for (const r of rows) {
        if (!books.has(r.playerId)) books.set(r.playerId, new Set());
        books.get(r.playerId)!.add(r.book);
        const key = r.market === 'anytime_td' ? 'anytime_td_prob' : r.market;
        const val = r.market === 'anytime_td' ? r.overProb : r.line;
        if (val == null) continue;
        if (!byPlayer.has(r.playerId)) byPlayer.set(r.playerId, new Map());
        const m = byPlayer.get(r.playerId)!;
        if (!m.has(key)) m.set(key, []);
        m.get(key)!.push(val);
    }
    const out = new Map<number, MarketProjection>();
    for (const [pid, markets] of byPlayer) {
        const values: Record<string, number> = {};
        for (const [m, vals] of markets) {
            const v = [...vals].sort((a, b) => a - b);
            const mid = v.length >> 1;
            values[m] = v.length % 2 ? v[mid] : (v[mid - 1] + v[mid]) / 2;
        }
        const [pts, used] = pprFromMarkets(values);
        out.set(pid, {
            values, pprPoints: pts, marketsPriced: used,
            booksPriced: books.get(pid)?.size ?? 0,
        });
    }
    return out;
}

export interface PropReport {
    status: 'ok' | 'skipped' | 'failed';
    reason?: string;
    events?: number; props?: number; players?: number;
}

export async function refreshPlayerProps(
    season: number, week: number, today: string,
): Promise<PropReport> {
    const key = (process.env.ODDS_API_KEY ?? '').trim();
    if (!key) {
        return { status: 'skipped', reason: 'ODDS_API_KEY not set' };
    }

    const pool = await query<{ id: number; full_name: string }>(
        `SELECT id, full_name FROM players
          WHERE redraft_pool = 1 AND position != 'DST'`);
    const seen = new Map<string, number[]>();
    for (const p of pool) {
        const k = normalizeName(p.full_name);
        (seen.get(k) ?? seen.set(k, []).get(k)!).push(p.id);
    }
    const byName = new Map<string, number>();
    for (const [k, v] of seen) if (v.length === 1) byName.set(k, v[0]);

    const get = async (url: string) => {
        const res = await fetch(url, { cache: 'no-store', signal: AbortSignal.timeout(45_000) });
        if (!res.ok) throw new Error(`${res.status} ${res.statusText}`);
        return res.json();
    };

    let events: any[];
    try {
        events = await get(`${API}/sports/${SPORT}/events?apiKey=${key}`);
    } catch (e) {
        return { status: 'failed', reason: String((e as Error)?.message ?? e) };
    }

    const markets = Object.keys(MARKETS).join(',');
    const rows: PropRow[] = [];
    for (const ev of events) {
        try {
            const odds = await get(
                `${API}/sports/${SPORT}/events/${ev.id}/odds`
                + `?apiKey=${key}&regions=us&oddsFormat=american&markets=${markets}`);
            rows.push(...parseEvent(odds, byName));
        } catch {
            // One event without a priced market is normal; the rest stand.
        }
    }

    const proj = consensus(rows);
    for (let i = 0; i < rows.length; i += 100) {
        const chunk = rows.slice(i, i + 100);
        const params: any[] = [];
        const values = chunk.map(r => {
            const b = params.length;
            params.push(r.playerId, season, week, r.market, r.book, r.line,
                r.overPrice, r.underPrice, r.overProb, r.eventId,
                r.commenceTime, today);
            return `(${Array.from({ length: 12 }, (_, k) => `$${b + k + 1}`).join(',')})`;
        });
        await getDb().prepare(
            `INSERT INTO nfl_player_prop
               (player_id, season, week, market, book, line, over_price,
                under_price, over_prob, event_id, commence_time, scraped_at)
             VALUES ${values.join(',')}
             ON CONFLICT (player_id, season, week, market, book, scraped_at)
             DO UPDATE SET line = excluded.line, over_price = excluded.over_price,
               under_price = excluded.under_price, over_prob = excluded.over_prob`,
        ).run(params);
    }

    const projRows = [...proj.entries()];
    for (let i = 0; i < projRows.length; i += 100) {
        const chunk = projRows.slice(i, i + 100);
        const params: any[] = [];
        const values = chunk.map(([pid, v]) => {
            const b = params.length;
            params.push(pid, season, week, v.values.receptions ?? null,
                v.values.rec_yds ?? null, v.values.rush_yds ?? null,
                v.values.pass_yds ?? null, v.values.pass_tds ?? null,
                v.values.rush_attempts ?? null, v.values.anytime_td_prob ?? null,
                v.pprPoints, v.marketsPriced, v.booksPriced, today);
            return `(${Array.from({ length: 14 }, (_, k) => `$${b + k + 1}`).join(',')})`;
        });
        await getDb().prepare(
            `INSERT INTO nfl_player_market_projection
               (player_id, season, week, receptions, rec_yards, rush_yards,
                pass_yards, pass_tds, rush_attempts, anytime_td_prob,
                ppr_points, markets_priced, books_priced, scraped_at)
             VALUES ${values.join(',')}
             ON CONFLICT (player_id, season, week, scraped_at)
             DO UPDATE SET receptions = excluded.receptions,
               rec_yards = excluded.rec_yards, rush_yards = excluded.rush_yards,
               pass_yards = excluded.pass_yards, pass_tds = excluded.pass_tds,
               rush_attempts = excluded.rush_attempts,
               anytime_td_prob = excluded.anytime_td_prob,
               ppr_points = excluded.ppr_points,
               markets_priced = excluded.markets_priced,
               books_priced = excluded.books_priced`,
        ).run(params);
    }

    return { status: 'ok', events: events.length, props: rows.length, players: proj.size };
}
